pkg/resolver/client/ens/ens_test.go

```diff
@@ -0,0 +1,149 @@
+// Copyright 2020 The Swarm Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package ens_test
+
+import (
+	"errors"
+	"testing"
+
+	"github.com/ethereum/go-ethereum/ethclient"
+	"github.com/ethersphere/bee/pkg/resolver/client/ens"
+	"github.com/ethersphere/bee/pkg/swarm"
+)
+
+func TestNewClient(t *testing.T) {
+	cl := ens.NewClient()
+	if cl.Endpoint != "" {
+		t.Errorf("expected no endpoint set")
+	}
+}
+
+func TestConnect(t *testing.T) {
+	ep := "test"
+
+	t.Run("no dial func error", func(t *testing.T) {
+		c := ens.NewClient(
+			ens.WithDialFunc(nil),
+		)
+		err := c.Connect(ep)
+		defer c.Close()
+		if err == nil && errors.Is(err, ens.ErrNotImplemented{}) {
+			t.Fatal("expected error")
+		}
+	})
+
+	t.Run("connect error", func(t *testing.T) {
+		c := ens.NewClient(
+			ens.WithErrorDialFunc(errors.New("failed to connect")),
+		)
+
+		if err := c.Connect("test"); err == nil {
+			t.Fatal("expected error")
+		}
+		c.Close()
+	})
+
+	t.Run("ok", func(t *testing.T) {
+		c := ens.NewClient(
```

Comment:
Defer a function that will close the client if it is still connected, as it will consume resources in case of an error in between the explicit Close call.

Reply:
Specifically, in this test, the Dial function is a noop. Calling Close() will cause a panic, as the ethereumClient underneath has not been properly initialized.

I will refactor the test in one of the future PRs to avoid this behavior.

Reply:
Wouldn't something like this work?

```go
defer func() {
    if c.IsConnected() {
        if err := c.Close(); err != nil {
            t.Fatal(err)
        }
    }
}()
```

Or maybe to create a resilient Close method, as it should not cause panic.

As in these tests, Close is called after the Connect call, but Close is usually associated with releasing resources acquired with the constructor, in this case NewClient. I would assume to always call Close when the instance is successfully created with the constructor.

Reply:
> Specifically, in this test, the Dial function is a noop. Calling Close() will cause a panic, as the ethereumClient underneath has not been properly initialized.

Then the `Close` function should have a nil check to prevent a panic